File: 2023/05 If You Give A Seed A Fertilizer/c/two/src/main.c
#include <ctype.h>
#include <stdio.h>
#include <string.h>

#include "utils.h"

typedef struct
{
    long long int start;
    long long int size;
} SeedRange;

/**
 * @brief Return true if the SeedRange pointed to by lhsVoid has a start less than rhsVoid's SeedRange
 * 
 * @param lhsVoid 
 * @param rhsVoid 
 * @return true 
 * @return false 
 */
static bool seedRangeLessThan(const void* const lhsVoid, const void* const rhsVoid)
{
    const SeedRange* const lhs = lhsVoid;
    const SeedRange* const rhs = rhsVoid;

    return lhs->start < rhs->start;
}

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        fprintf(stderr, "Usage: %s <INPUT_FILE>\n", argv[0]);
        return 1;
    }

    FILE* const infile = fopen(argv[1], "r");
    if (infile == NULL)
    {
        fprintf(stderr, "ERROR! Failed to open \"%s\"\n", argv[1]);
        return 1;
    }

    // Verify start of seed line
    int bytesRead = 0;
    fscanf(infile, "seeds: %n", &bytesRead);
    if (bytesRead != 7)
    {
        fputs("ERROR! Failed to read seeds list\n", stderr);
        fclose(infile);
        return 1;
    }

    // Used to make cleanup easier
    int exitStatus = 0;

    DynamicArray seeds;
    dynamicArrayCreate(&seeds, sizeof(SeedRange));

    // Read seed ranges
    for (long long int start, size; fscanf(infile, "%lld %lld", &start, &size) == 2;)
    {
        const SeedRange range = {.start = start, .size = size};
        dynamicArrayPushBack(&seeds, &range, 1);
    }

    // Array of bool flags representing if a seed range at a given index was covnerted during the current map
    // Reset at start of each map
    DynamicArray rangeConverted;
    dynamicArrayCreate(&rangeConverted, sizeof(unsigned char));
    dynamicArrayResize(&rangeConverted, seeds.size);

    // New breakaway ranges to insert into seeds after each seed is converted. Cleared after insertion
    DynamicArray newSeeds;
    dynamicArrayCreate(&newSeeds, sizeof(SeedRange));

    // Read each line

    LineBuffer line;
    lineBufferCreate(&line);

    for (int lineNum = 1; readNextLine(infile, &line); ++lineNum)
    {
        // Skip empty lines
        if (line.size == 0)
            continue;

        // Process mapping if line starts with digit
        if (isdigit(line.line[0]))
        {
            // Read mapping
            long long int destStart;
            long long int srcStart;
            long long int size;
            if (sscanf(line.line, "%lld %lld %lld", &destStart, &srcStart, &size) != 3)
            {
                fprintf(stderr, "ERROR! Failed to read line %d\n", lineNum);
                exitStatus = 1;
                goto cleanup;
            }

            // Try to convert all seeds
            for (size_t i = 0; i < seeds.size; ++i)
            {
                unsigned char* const converted = (unsigned char*)dynamicArrayIndex(&rangeConverted, i);
                if (*converted)
                    continue;

                SeedRange* const startingRange = (SeedRange*)dynamicArrayIndex(&seeds, i);

                // If any part of this range should be converted
                if (startingRange->start + startingRange->size - 1 >= srcStart && startingRange->start < srcStart + size)
                {
                    // Break off portion before start of conversion area
                    if (startingRange->start < srcStart)
                    {
                        const SeedRange newRange = {.start = startingRange->start, srcStart - startingRange->start};
                        dynamicArrayPushBack(&newSeeds, &newRange, 1);

                        startingRange->size -= srcStart - startingRange->start;
                        startingRange->start = srcStart;
                    }

                    // Break off portion before after end of conversion area
                    if (startingRange->start + startingRange->size > srcStart + size)
                    {
                        const SeedRange newRange = {.start = srcStart + size, startingRange->start + startingRange->size - (srcStart + size)};
                        dynamicArrayPushBack(&newSeeds, &newRange, 1);

                        startingRange->size -= startingRange->start + startingRange->size - (srcStart + size);
                    }

                    // Convert remaining portion
                    startingRange->start = destStart + startingRange->start - srcStart;
                    *converted = 1;

                    // Insert any new seed ranges from above breakaway code
                    dynamicArrayPushBack(&seeds, newSeeds.array, newSeeds.size);
                    dynamicArrayClear(&newSeeds);

                    // Resize to match current seed count
                    dynamicArrayResize(&rangeConverted, seeds.size);
                }
            }
        }
        // Otherwise a new mapping is starting. Reset conversion flags
        else
        {
            memset(rangeConverted.array, 0, sizeof(unsigned char) * rangeConverted.size);
        }
    }

    printf("Result: %lld\n", ((const SeedRange*)min(seeds.array, sizeof(SeedRange), seeds.size, seedRangeLessThan))->start);

cleanup:
    lineBufferDestroy(&line);

    dynamicArrayDestroy(&newSeeds);
    dynamicArrayDestroy(&rangeConverted);
    dynamicArrayDestroy(&seeds);

    fclose(infile);

    return exitStatus;
}

File: 2023/05 If You Give A Seed A Fertilizer/rust/two/src/main.rs
use std::env;
use std::fs;

#[derive(Debug)]
enum ParseSeedsError {
    EmptyLine,
    OddNumberCount,
    WrongStart,
}

struct SeedRange {
    start: i64,
    size: i64,
}

impl SeedRange {
    fn from(start: i64, size: i64) -> SeedRange {
        SeedRange { start, size }
    }
}

// Parses seed ranges from string in format of "seeds: <RANGE_START> <RANGE_END> <RANGE_START> <RANGE_END>"
fn parse_seeds(string: &str) -> Result<Vec<SeedRange>, ParseSeedsError> {
    let mut tokens = string.split(" ");

    // Make sure at least one token is present
    if let Some(start) = tokens.next() {
        // Verify start of line
        if start == "seeds:" {
            let mut seeds = Vec::new();

            // Parse all found seeds and insert them into seeds
            while let Some(seed) = tokens.next() {
                let seed = seed
                    .trim()
                    .parse()
                    .expect("Unexpcted non-number in seeds list");
                seeds.push(seed);
            }

            // Make sure all seeds have both start and size
            if seeds.len() % 2 == 1 {
                Err(ParseSeedsError::OddNumberCount)
            } else {
                Ok(seeds
                    .chunks(2)
                    .map(|chunk| SeedRange::from(chunk[0], chunk[1]))
                    .collect())
            }
        } else {
            Err(ParseSeedsError::WrongStart)
        }
    } else {
        Err(ParseSeedsError::EmptyLine)
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Usage: {} <INPUT_FILE>", args[0]);
    }

    // Read file into lines
    let contents = fs::read_to_string(&args[1]).expect("Failed to read input file");
    let lines: Vec<&str> = contents.lines().collect();

    let mut seeds = parse_seeds(lines[0]).expect("Failed to parse seeds");

    // New breakaway ranges to insert into seeds after each seed is converted. Cleared after insertion
    let mut new_seeds = Vec::new();

    // Flag to check if a range at a given index has been converted yet by the current map. All set to false on start of every map
    let mut range_converted: Vec<bool> = vec![false; seeds.len()];

    // Iterate over each line, skipping seed line, and removing empty lines
    for (line_num, line) in lines
        .iter()
        .enumerate()
        .skip(2)
        .filter(|(_line_num, line)| !line.is_empty())
    {
        // If line starts with digit it's a mapping. Otherwise it's the start of a new map
        if line.chars().nth(0).unwrap().is_ascii_digit() {
            // Parse numbers on line
            let tokens: Vec<i64> = line
                .split(" ")
                .map(|string| {
                    string
                        .trim()
                        .parse::<i64>()
                        .expect("Failed to parse mapping value")
                })
                .collect();

            // There should only be three numbers per mapping
            if tokens.len() != 3 {
                panic!("Invalid number on line {line_num}");
            }

            // Extract three numbers by name
            let dest_start = tokens[0];
            let src_start = tokens[1];
            let size = tokens[2];

            // Iterate by index because we need to add ranges mid iteration
            for i in 0..seeds.len() {
                let converted = &mut range_converted[i];

                // Skip ranges already converted during current map
                if *converted {
                    continue;
                }

                let range = &mut seeds[i];

                // Skip ranges that don't overlap with the conversion area
                if range.start + range.size - 1 < src_start || range.start >= src_start + size {
                    continue;
                }

                // Break off portion before start of conversion area
                if range.start < src_start {
                    new_seeds.push(SeedRange::from(range.start, src_start - range.start));
                    range.size -= src_start - range.start;
                    range.start = src_start;
                }

                // Break off portion before after end of conversion area
                if range.start + range.size > src_start + size {
                    new_seeds.push(SeedRange::from(
                        src_start + size,
                        range.start + range.size - (src_start + size),
                    ));
                    range.size -= range.start + range.size - (src_start + size);
                }

                // Convert remaining portion
                range.start = dest_start + range.start - src_start;
                *converted = true;

                // Insert breakaway seeds
                seeds.append(&mut new_seeds);

                // Resize to match current seed count
                range_converted.resize(seeds.len(), false);
            }
        } else {
            // Reset conversion flags for start of new map
            range_converted.fill(false);
        }
    }

    println!(
        "Result: {}",
        seeds
            .iter()
            .map(|range| range.start)
            .min()
            .expect("No seed ranges found")
    );
}
